File: Client/src/modules/models/spell/FireballSpell.js
var FireballSpell = Spell.extend({
    ctor:function(){
        //Spell.prototype.ctor(3).call();
        //this.init(3);
        this._super(0);
        this.effectTick = 1;
        this.range = 0.8;
        this.leftTick = 1;
        this.animTick = 30;
    },
    setUI:function (spellUI){
        this.ui = spellUI;
    },
    update:function (monsters, towers){
        cc.log("FireballSpell update:" + this.leftTick + ", curpos: " + this.curPos.x + " " + this.curPos.y);
        if(this.leftTick > 0){
            if(this.animTick > 0){
                this.animTick--;
            }
            else{
                for(let i = 0; i < monsters.length; ++i){
                    let monster = monsters[i];
                    if (Helper.getDistance(this.curPos, monster.curPos) <= this.range && monster.isBuffBySpell(this) == false) {
                        monster.addSpell(this);
                        this.doEffect(monster);
                    }
                }
                this.leftTick -= 1;
            }

            return true;
        }
        else {
            if(this.ui){
                this.ui.removeFromParent();
            }

            return false;
        }
    },
    removeEffect:function (object){

    },
    doAction:function (object){

    },
    setBattleId:function (battleId){
        this.battleId = battleId;
    },
    doEffect:function (monster){
        monster.hitBullet(50);
    },
    initAnim:function (){
        let path = this.animPath;
        this.anim = new sp.SkeletonAnimation(path + ".json", path + ".atlas");
        this.anim.setAnchorPoint(cc.p(0.5, 0.5));
        this.anim.setPosition(cc.p(0, 100));
        //this.anim.setAnimation(0, "animation_ice_ball", false);
        this.anim.addAnimation(0,"animation_fireball", false, 0);
        this.anim.addAnimation(0,"animation_full", true, 0.3);
        this.anim.update(0.0);
        let action = cc.MoveBy.create(0.3, cc.p(0, -100));
        this.anim.runAction(action);
        /*this.anim.scheduleOnce(function(){
            this.setAnimation(0, "animation_bottom", true);
        }, 0.5);*/
        return this.anim;
    }
});

File: Server/src/model/Spell/FireballSpell.java
package model.Spell;

import model.Battle.BattleObject;
import model.Define.ModelDefine;
import model.Monster.Monster;
import model.Tower.Tower;

import java.util.Iterator;
import java.util.LinkedList;

public class FireballSpell extends Spell{
    private int animTick;
    public FireballSpell(int level) {
        super(0, 1);
        readSpellJson();
    }
    protected void readSpellJson(){
        super.readSpellJson();
        this.effectTick = 1;
        this.leftTick = 1;
        this.animTick = 30;
    }
    public boolean update(LinkedList<Monster> monsters, LinkedList<Tower>towers){
        System.out.println("Spell Fireball" + battleObjectId + " time left: " + leftTick);
        if(leftTick > 0){
            if(this.animTick > 0){
                this.animTick--;
            }
            else{
                Iterator<Monster> itr = monsters.iterator();
                while(itr.hasNext()) {
                    Monster monster = itr.next();
                    if (curPos.getDistance(monster.getCurPos()) <= range && monster.isBuffBySpell(this) == false) {
                        monster.addSpell(this);
                        doEffect(monster);
                    }
                }
                leftTick -= 1;
            }
            return true;
        }
        else {
            return false;
        }
    }

    private void doEffect(Monster monster) {
        monster.hit(50);
    }

    public void doAction(BattleObject battleObject){

    }

    public void removeEffect(BattleObject battleObject){
    }

    @Override
    public int getAnimTick() {
        return animTick;
    }
}
